File: metodi-di-pagamento-alternativi/amazonpay/python/amazonpay.py
# -*- coding: utf-8 -*-

# Amazon Pay - Pagamenti integrati

import sys
if sys.version_info >= (3,):
    from urllib.parse import urlencode
else:
    from urllib import urlencode
import hashlib
import datetime
import time
import requests

HTTP_HOST = "my-server.example.tdl"

requestUrl = "https://int-ecommerce.nexi.it/ecomm/api/paga/amazonpay"

# Alias e chiave segreta
APIKEY = "<ALIAS>" # Sostituire con il valore fornito da Nexi
CHIAVESEGRETA = "<CHIAVE SEGRETA PER CALCOLO MAC>" # Sostituire con il valore fornito da Nexi

# Parametri per calcolo MAC
codTrans = "TESTAP_" + datetime.datetime.today().strftime('%Y%m%d%H%M%s')
importo = 5000
divisa = "978"
timeStamp = int(time.time()*1000)

# Calcolo MAC
mac_str = 'apiKey=' + str(APIKEY) + 'codiceTransazione=' + str(codTrans)  + 'importo=' + str(importo) + 'divisa=' + str(divisa) + "timeStamp=" + str(timeStamp) + str(CHIAVESEGRETA)
mac =  hashlib.sha1(mac_str.encode('utf8')).hexdigest()

# Parametri dati
dati = {
    'apiKey': APIKEY,
    'codiceTransazione': codTrans,
    'importo': importo,
    'divisa': divisa,
    'amazonpay': {
        'amazonReferenceId': '',
        'accessToken': '',
        'softDecline': '',
        'creaContratto': ''
    },
    'parametriAggiuntivi': {
        'nome': 'Mario',
        'cognome': 'Rossi',
        'mail': "cardHolder@mail.it",
        'descrizione': "descrizione",
        'Note1': "note",
    },
    'timeStamp': str(timeStamp),
    'mac': mac,
}

response = requests.post(requestUrl, json=dati)

dataVerifica = response.json()

if dataVerifica['esito'] == "OK":
    macCalculated_str = 'esito=' + dataVerifica['esito'] + 'idOperazione=' + dataVerifica['idOperazione'] + 'timeStamp=' + dataVerifica['timeStamp'] + CHIAVESEGRETA
    macCalculated =  hashlib.sha1(macCalculated_str.encode('utf8')).hexdigest()
    if macCalculated != dataVerifica['mac']:
        print('Errore MAC: ' + macCalculated + ' non corrisponde a ' + dataVerifica['mac'])
    else:
        print('La transazione ' + codTrans + " è avvenuta con successo; codice autorizzazione: " + dataVerifica['codiceAutorizzazione'])
else:
    print('La transazione ' + codTrans + " è stata rifiutata; descrizione errore: " + dataVerifica['errore']['messaggio'])
                    
                

File: metodi-di-pagamento-alternativi/amazonpay/java/amazonpay.java
// Amazon Pay - Pagamenti integrati

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.json.JSONObject;

import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.Date;

public class Esempio {

    public static void main(String[] a) throws Exception {

        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss");
        Date date = new Date();

        // Parametri per calcolo MAC
        String apiKey = "<ALIAS>"; // Alias fornito da Nexi
        String chiaveSegreta = "<CHIAVE SEGRETA PER CALCOLO MAC>"; // Chiave segreta fornita da Nexi

        String requestUrl = "https://int-ecommerce.nexi.it/" + "ecomm/api/paga/amazonpay";

        String codTrans = "TESTAP_" + dateFormat; // Codice della transazione
        String importo = "5000"; // 5000 = 50,00 EURO (indicare la cifra in centesimi)
        String divisa = "978"; // divisa 978 indica EUR
        String timeStamp = "" + System.currentTimeMillis();
        
        JSONObject jsonAmazonpay = new JSONObject();
        jsonAmazonpay.put("amazonReferenceId", "");
        jsonAmazonpay.put("accessToken", "");
        jsonAmazonpay.put("softDecline", "");
        jsonAmazonpay.put("creaContratto", "");    

        String stringaMac = "apiKey=" + apiKey
          + "codiceTransazione=" + codTrans
          + "importo=" + importo
          + "divisa=" + divisa
          + "timeStamp=" + timeStamp
          + chiaveSegreta;

        // Calcolo MAC
        String macCalculated = hashMac(stringaMac);

        // Parametri di invio
        JSONObject json = new JSONObject();
        json.put("apiKey", apiKey);
        json.put("codiceTransazione", codTrans);
        json.put("importo", importo);
        json.put("divisa", divisa);
        json.put("amazonpay", jsonAmazonpay);
        json.put("timeStamp", timeStamp);
        json.put("mac", macCalculated);
        
        JSONObject parametriAggiuntivi = new JSONObject();
        
        parametriAggiuntivi.put("nome", "Mario");
        parametriAggiuntivi.put("cognome", "Rossi");
        parametriAggiuntivi.put("mail", "cardHolder@mail.it");
        parametriAggiuntivi.put("descrizione", "descrizione");
        parametriAggiuntivi.put("Note1", "note");

        json.put("parametriAggiuntivi", parametriAggiuntivi);

        // Chiamata API
        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
        HttpPost request = new HttpPost(requestUrl);
        StringEntity params = new StringEntity(json.toString());
        request.addHeader("content-type", "application/json");
        request.setEntity(params);
        HttpResponse response = httpClient.execute(request);
        HttpEntity entity = response.getEntity();
        String responseString = EntityUtils.toString(entity, "UTF-8");

        // Parametri di ritorno
        JSONObject responseObj = new JSONObject(responseString);

        System.out.print(responseObj.toString());
    }

    public static String hashMac(String stringaMac) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-1");
        byte[] in = digest.digest(stringaMac.getBytes("UTF-8"));

        final StringBuilder builder = new StringBuilder();
        for (byte b : in) {
            builder.append(String.format("%02x", b));
        }
        return builder.toString();
    }
}
                    
                